(fix): allow title changes that don't modify filename

herbertjones opened the pull request:
- Prevent error when trying to move file to same name
- Prevent string matching in org from messing with the current re-search-forward
  position which has caused infinite broken search and replace in testing (might
  not affect all org versions or may be due to certain code paths)

###### Motivation for this change

I tried to change the title of a page from "Webassembly" to "WebAssembly", which caused an error due to renaming the file to the existing name.  After fixing that, auto-updating links began corrupting files in this case.  A file with a link to "Webassembly" began infinitely inserting (broken) links to the "webassembly.org" file at the beginning of the file, as something was evidently changing the match position and causing `replace-match` in `org-roam--replace-link` to replace the wrong area.

d12frosted commented:
Just curious, are you on macos?

herbertjones commented:
@d12frosted No, I'm on Debian Linux testing (Bullseye).

jethrokuan commented on org-roam.el:
```diff
@@ -1302,21 +1302,24 @@ If OLD-DESC is passed, and is not the same as the link
-                                  new-desc
-                                label)))
-              (replace-match (org-roam-format-link new-path new-label type)))))))))
+  (cl-flet ((get-new-link ()
```
I don't think we need a cl-flet here. I think `org-in-regexp` butchers the old match data. Simpler to just wrap the `when-let` clause in a `save-match-data`, without introducing it as a function.

herbertjones commented:
Problem definitely occurs if I don't wrap `(link (org-element-lineage (org-element-context) '(link) t))` also.   Looks like `org-roam--fix-relative-links` may need to be fixed too possibly.

jethrokuan commented:
Have you tested that this works for you? What I meant was to `save-match-data` for everything within the original `when-let` clause, which is essentially what the original PR did, except with an additional named function. Now it's just saving for `org-element-lineage`.

herbertjones commented:
Yeah, I had both branches wrapped with `save-match-data` as they both needed it to work without erroring.  I've moved it to a separate named function in the change.  Also, since it was most likely a problem I updated the other possible problem loop similarly.
